numpy/core/src/multiarray/methods.c

```diff
@@ -1008,6 +1008,44 @@ array_getarray(PyArrayObject *self, PyObject *args)
 
 
 static PyObject *
+array_ufunc(PyArrayObject *self, PyObject *args, PyObject *kwds)
+{
+    PyObject *ufunc, *method_name, *normal_args, *ufunc_method, *result;
+
+    if (PyTuple_Size(args) < 2) {
+        PyErr_SetString(PyExc_TypeError,
+                        "__array_ufunc__ requires at least 2 arguments");
+        return NULL;
+    }
+    ufunc = PyTuple_GET_ITEM(args, 0);  /* borrowed reference */
+    if (ufunc == NULL) {
+        return NULL;
+    }
+
+    method_name = PyTuple_GET_ITEM(args, 1);
+    if (method_name == NULL) {
+        return NULL;
+    }
+
+    normal_args = PyTuple_GetSlice(args, 2, PyTuple_GET_SIZE(args));
+    if (normal_args == NULL) {
+        return NULL;
+    }
+
+    ufunc_method = PyObject_GetAttr(ufunc, method_name);
+    if (ufunc_method == NULL) {
+        Py_DECREF(normal_args);
+        return NULL;
+    }
+
+    result = PyObject_Call(ufunc_method, normal_args, kwds);
```

Comment:
@charris - `ndarray.__array_ufunc__` never gets called by the `ufunc` directly, only by a subclass that does `result = super(subclass, self).__array_ufunc__(*inputs, **kwargs)`. In this use, it simply replaces the `getattr(ufunc, method)(*inputs, **kwargs)` that was initially recommended (indeed, as is, it just does that; this can be sped up later). The main advantage of all this is that by making super possible, we *allow* multiple inheritance using the standard python method

@njsmith - for context, maybe most useful to look at the updated docs, in particular https://github.com/charris/numpy/pull/5/files#diff-5a79b101110fd040a0cd5a254b4742f6R515 (if that isn't clear, let me know -- it is supposed to explain why we do things this way!).